src/lib/shmem/src/allocator.rs

```diff
@@ -0,0 +1,252 @@
+                c_bindings::shmemblockserialized_fromString(buf.as_ptr(), &mut err)
+            },
+        };
+        if err {
+            None
+        } else {
+            Some(res)
+        }
+    }
+}
+
+pub struct Allocator {
+    internal: *mut c_bindings::ShMemAllocator,
+}
+
+// SAFETY: The C bindings for ShMemAllocator use internal locking to ensure
+// thread-safe access.
+unsafe impl Send for Allocator {}
+unsafe impl Sync for Allocator {}
+
+impl Allocator {
+    pub fn global() -> &'static Self {
+        static GLOBAL: OnceCell<Allocator> = OnceCell::new();
+        GLOBAL.get_or_init(|| Self {
+            internal: unsafe { c_bindings::shmemallocator_getGlobal() },
+        })
+    }
+
+    pub fn alloc<T>(&self, val: T) -> ShMemBlock<T>
+    where
+        T: Sync + VirtualAddressSpaceIndependent,
+    {
+        let nbytes = std::mem::size_of_val(&val);
+        let raw_block =
+            unsafe { c_bindings::shmemallocator_alloc(self.internal, nbytes.try_into().unwrap()) };
+        unsafe { (raw_block.p as *mut T).write(val) };
+        ShMemBlock {
+            internal: raw_block,
+            allocator_or_serializer: AllocatorOrSerializer::Allocator(self),
+            _phantom: PhantomData::<T>::default(),
+        }
+    }
+
+    /// Deserialize a `block` that was allocated by `self`. You probably want
+    /// Serializer::deserialize instead.
+    ///
+    /// TODO: do we need this? Can only be called from the process
+    /// that originally allocated the block, which shouldn't need to.
+    ///
+    /// SAFETY:
+    /// * `block` must have been originally created by this allocator.
+    /// * `block` must have been created from  a value of type `T`.
+    /// * The backing memory must be live for the lifetime of the returned
+    ///   ShMemBlock. e.g. `ShMemBlock::deallocate` must not have been called
+    ///   on any block referencing that memory.
+    pub unsafe fn deserialize<'a, T>(&'a self, block: &ShMemBlockSerialized) -> ShMemBlock<'a, T>
+    where
+        T: Sync + VirtualAddressSpaceIndependent,
+    {
+        ShMemBlock {
+            internal: unsafe {
+                c_bindings::shmemallocator_blockDeserialize(self.internal, &block.internal)
+            },
+            allocator_or_serializer: AllocatorOrSerializer::Allocator(&self),
+            _phantom: PhantomData::default(),
+        }
+    }
+}
+
+pub struct Serializer {
+    internal: *mut crate::c_bindings::ShMemSerializer,
+}
+
+// SAFETY: The C bindings for ShMemSerializer use internal locking to ensure
+// thread-safe access.
+unsafe impl Send for Serializer {}
+unsafe impl Sync for Serializer {}
+
+impl Serializer {
+    pub fn global() -> &'static Self {
+        static GLOBAL: OnceCell<Serializer> = OnceCell::new();
+        GLOBAL.get_or_init(|| Self {
+            internal: unsafe { c_bindings::shmemserializer_getGlobal() },
+        })
+    }
+
+    /// SAFETY:
+    /// * `self` must have been created from a `ShMemBlock<T>`
```

Comment:
What does this mean? Isn't `self` (a `&Serializer`) created by `Serializer::global()`?

Reply:
Oops; missed updating this comment when I moved this method. (Was originally a static method of ShMemBlock)